sync-core/src/test/java/com/cloudant/http/HttpTest.java

```diff
@@ -119,4 +122,90 @@ public void testReadBeforeExecute() throws IOException {
+    // be named cookie_test
+    //
+    @Test
+    public void testCookieAuthWithoutRetry() throws IOException {
+
+        if(IGNORE_AUTH_HEADERS){
+            System.err.println("Ignore cookie filters");
+            return;
+        }
+
+        CookieFilter filter = new CookieFilter(System.getProperty("test.couch.username"),System.getProperty("test.couch.password"));
+
+        CouchConfig config = getCouchConfig("cookie_test");
+        HttpConnection conn = new HttpConnection("POST", config.getRootUri().toURL(),
+                "application/json");
+        conn.responseFilters.add(filter);
+        conn.requestFilters.add(filter);
+        ByteArrayInputStream bis = new ByteArrayInputStream(data.getBytes());
+
+        // nothing read from stream
+        Assert.assertEquals(bis.available(), data.getBytes().length);
+
+        conn.setRequestBody(bis);
+        conn.execute();
+
+        // stream was read to end
+        Assert.assertEquals(bis.available(), 0);
+        Assert.assertEquals(2, conn.getResponseCode() / 100);
+
+        //check the json
+        JSONHelper helper = new JSONHelper();
+        Map<String,Object> jsonRes = helper.fromJson(new InputStreamReader(conn.getConnection()
+                .getInputStream()));
+
+        Assert.assertTrue(jsonRes.containsKey("ok"));
+        Assert.assertTrue((Boolean)jsonRes.get("ok"));
+        Assert.assertTrue(jsonRes.containsKey("id"));
+        Assert.assertTrue(jsonRes.containsKey("rev"));
+
+    }
+
+    @Test
+    public void testCookieAuthWithRetry() throws IOException {
+
+        if(IGNORE_AUTH_HEADERS){
+            System.err.println("Ignore cookie filters");
+            return;
+        }
+
+        CookieFilter filter = new CookieFilter(System.getProperty("test.couch.username"),System.getProperty("test.couch.password"));
+
+        CouchConfig config = getCouchConfig("cookie_test");
+        CouchClient client = new CouchClient(config);
+        HttpConnection conn = new HttpConnection("POST", config.getRootUri().toURL(),
+                "application/json");
+        conn.responseFilters.add(filter);
+        ByteArrayInputStream bis = new ByteArrayInputStream(data.getBytes());
+
+        // nothing read from stream
+        Assert.assertEquals(bis.available(), data.getBytes().length);
+
+        conn.setRequestBody(bis);
+        conn.execute();
+
+        // stream was read to end
+        Assert.assertEquals(bis.available(), 0);
+        Assert.assertEquals(2, conn.getResponseCode() / 100);
+
+        //check the json
+        JSONHelper helper = new JSONHelper();
+        Map<String,Object> jsonRes = helper.fromJson(new InputStreamReader(conn.getConnection()
+                .getInputStream()));
+
+        Assert.assertTrue(jsonRes.containsKey("ok"));
+        Assert.assertTrue((Boolean)jsonRes.get("ok"));
+        Assert.assertTrue(jsonRes.containsKey("id"));
+        Assert.assertTrue(jsonRes.containsKey("rev"));
```

Comment:
I don't think this should succeed because you've not set a request filter. The response filter shouldn't be able to modify future (retried) requests, only respond to or alter the results of the current one.